File: docs/JAMIA/comparisons/RPython/TM/37513355_spark_mixed_types.py
# ID: SO#37513355
# Language: Python
# Bug class: Data loading
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: Spark schema inference fails on mixed types. Load-time type inference error.
# Reproduction status: Live
# Type system catch: Schema rejects mixed-type columns; `number | null` not `number | string`

import pandas as pd
import numpy as np

# Simulate the SO scenario: CSV data where a column has mixed types
# (some values look numeric, others are strings)
df = pd.DataFrame({
    "id": ["10000001", "10000001", "10000002"],
    "status": ["OK", "OK", "PA"],
    "score": [543, 611, 691],
    "mixed_col": [1.5, "NA", 3.2],  # This is the problem column
})

print("DataFrame dtypes:")
print(df.dtypes)
print()
print("mixed_col dtype is 'object' because it contains both float and str:")
print(f"  dtype: {df['mixed_col'].dtype}")
print(f"  values: {df['mixed_col'].tolist()}")
print()

# The crash would happen here with PySpark:
# from pyspark.sql import SparkSession
# spark = SparkSession.builder.getOrCreate()
# sdf = spark.createDataFrame(df)
# TypeError: Can not merge type <class 'pyspark.sql.types.StringType'>
#            and <class 'pyspark.sql.types.DoubleType'>

# Without PySpark, demonstrate that pandas silently allows the mixed column:
print("pandas allows this silently — the column is object dtype.")
print("Spark would crash when trying to infer a consistent schema.")
print()

# Attempting numeric operations on the mixed column shows the issue:
try:
    result = df["mixed_col"].astype(float)
except (ValueError, TypeError) as e:
    print(f"Converting to float crashes: {e}")

File: docs/JAMIA/comparisons/RPython/TM/37513355_spark_mixed_types.ts
/**
 * RPython SO#37513355 — Converting Pandas DataFrame into Spark DataFrame error
 * Effect: Crash
 * Bug class: Type coercion
 *
 * Python bug: A DataFrame has a column with mixed float and string values:
 * `mixed_col: [1.5, "NA", 3.2]`. Pandas allows this (object dtype), but Spark's
 * createDataFrame crashes: "Can not merge type StringType and DoubleType".
 *
 * In tidy-ts, a column with mixed types is typed as the union (number | string).
 * Downstream numeric operations on that column are rejected at compile time.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";

// Same data as the .py: most columns are fine, but mixed_col has both numbers and strings
const df = createDataFrame([
  { id: "10000001", status: "OK", score: 543, mixed_col: 1.5 as number | string },
  { id: "10000001", status: "OK", score: 611, mixed_col: "NA" as number | string },
  { id: "10000002", status: "PA", score: 691, mixed_col: 3.2 as number | string },
]);

// The user wants to do numeric operations on mixed_col (e.g., compute mean).
// In pandas, object dtype silently allows this. In Spark, it crashes on schema.
// In tidy-ts, s.mean() rejects (number | string)[] at compile time.
// @ts-expect-error — (number | string)[] is not assignable to number[]
s.mean(df.extract("mixed_col"));
